### ISSUE TITLE
PLIST Rotation issues

### ISSUE THREAD

**Issue by peardox** [mtgjson/mtgjson#766]
First I'll mention the obvious, this one is admittedly of limited interest (basically important only to those, like myself, who wanna know the source of a given printing)

With the release of [KHM](https://mtg.fandom.com/wiki/Kaldheim) WotC rotated 48 cards in [PLIST](https://mtg.fandom.com/wiki/The_List) removing some from the [ZNR](https://mtg.fandom.com/wiki/Zendikar_Rising) version of PLIST.

Unfortunately, Scryfall don't appear to make any distinction between PLISTZNR and PLISTKHM

We're bound to get a PLISTSTX etc going forward for a while all of which will be in PLIST.

Fortunately, the Fandom Wiki (link above) has a handy table that details the rotation. A quick little test shows that it cuts & pastes as a Tab-Separated variant of CSV (attached) with the following layout...

SetCode
SetName
CardName (needs tidying-up)
ZNRInclusion
KHMInclusion

This is all readily machine-processable (I'm about to play with it)

One issue is that without an official machine-friendly source Fandom has to be taken at face value to be 100% correct (I'll assume it is). Grabbing the Wiki source and processing that probably makes sense longer-term, but for now, the CSV will suffice.

For the moment, personally, I'm just going to create children of PLIST for the set-rotation of this set so everything has a proper home

[plist.txt](https://github.com/mtgjson/mtgjson/files/5974154/plist.txt)

**Comment by peardox** [comment]
For anyone who's interested I've split out PLIST into PLIST (Cards in both versions), PLISTZNR (ZNR only) and PLISTKHM (KHM only)

Note that I'm only providing the details of where a card 'properly' lives, you'll need to apply it yourself to your own setup (unless MTGJson decide to do something similar to my version).

Note that this also breaks SetList, and PLIST a bit as PLIST now only has 252 cards while PLISTZNR + PLISTKHM have 48 mutually exclusive sets that add to the specific set.

The easiest way to let you have the corrected lists is as a simple JSON 'errata' file which is now available if you want it on https://decknet.co.uk/fixups.php

The structure is simply two setlists in one file - it's very simple, examine it and it's obvious what it's doing. Each record has the 4 field - e.g...

```
{
    "uuid": "3618b02b-65c6-539d-91b5-6c3b65c9fb6b",
    "cardname": "Brazen Borrower \/\/ Petty Theft",
    "shortname": "Brazen Borrower",
    "setcode": "PLISTZNR"
}

```

Yeah, the names are superfluous - there're so you (and I) can check things - just as well as I almost missed a few with multiple face records, the above being an example (it's got two uuids)

Basically, wherever you have a uuid in the json list it can optionally be assigned to a new child set of PLIST

If you wanna see what the sets look like when applied I've for a dev page set up ...

https://decknet.co.uk/?set=plist&ord=price&prov=ck
https://decknet.co.uk/?set=plistznr&ord=price&prov=ck
https://decknet.co.uk/?set=plistkhm&ord=price&prov=ck

All this stuff is updated daily about an hour after MTGJson updates but realistically this stuff won't change until STX (I've already built the capacity into the build process - when the list gets updated I just need to update the csv)

**Comment by kodawah** [comment]
this is interesting, and very much needed.

design-wise i'd just avoid introducing a new set code, but rather keep the "official" PLIST code and add a subfield for the parent code. Alternatively the PLIST card objects could have an array similar to `Printings` containing all the associated sets whose PLIST card appears in (it's not excluded that a card may be added, then removed, then added again).

would you be able to provide the code that scrapes mtgwiki data?

**Comment by kodawah** [comment]
the same should/could be done for the silly promo pack cards btw

**Comment by peardox** [comment]
I made https://decknet.co.uk/fixups.php deliberately as structure-neutral as possible as that way the data is easy to adapt to whatever structure you want to use personally for your data. The fixups json only says that a specific card uuid is part of my virtual PLISTZNR/PLISTKHM  sets. Armed with that info any dev can implement their own solution such as your Printings concept.

If you look at https://scryfall.com/sets (scryfall being a primary MTGJson data source) then you'll note that all sets form a tree, PLIST is a supplementary set (which can optionally be added to a main set). My own use scenario is that I want to know what cards are potentially in a booster so the tree makes sense for myself (each to their own)

As for the Wiki stuff I just used a cut&paste from the page as that was easy to decode. Having said that there are a few gotchas using such free-form data (several non-standard namings for cards, having to strip off Wiki note markers, outright renaming cards). There is also the case of DFCs potentially being missed (did that while experimenting)

All that said I will be looking at using a proper Wiki page extraction as that appears to allow for the removal problem text (links being the main one). I'm about to do a Wiki comparison for #765 and that indicates that a Wiki export is preferable in the future.

With regard to the silly promo pack cards PKHM just came out today I notice. Or are you talking about the not-yet-on-scryfall mini-game (MKHM?) and art (AKHM?) cards - I just purchased a KHM set booster specifically to see physial cards).

If AKHM + MKHM I'm looking into doing those ATM before Scryfall get around to it as a stop-gap. I've also ordered full sets of AZNR + AKHM (they're cheap) so I can scan the backs - they're mostly blank but there is some text - scryfall just has 'no image available' images. I'm de-cyphering how @Zach creates uuids to make mine at least MTGJson friendly.

As for code, as mentioned above I'm not actually scraping any Wiki yet but a when I do I'll be starting off by reading https://www.mediawiki.org/wiki/API:Get_the_contents_of_a_page properly as this shuld, theoretically, also work on Fandom (they also use MediaWiki I believe - I'll know for sure when I try it out)

The code I have is PHP and uses a lot of MySQL - people around here seem to like Python

If you ever do any research into AI the web is littered with some bloody awful Python - put me off even thinking about learning the language...

**Comment by ZeldaZach** [comment]
I like the concept of multiple PLIST files personally, each with their own annotation on which they're based on. Will bring up this idea with an upstream provider and get their take.

**Comment by peardox** [comment]
This conceptually applies to other sets - JMP springs to mind as they have replacement cards for Arena which muddles things up. I'm sure Arena will happily trample over other sets when it feels like it...

**Comment by kodawah** [comment]
imo it's a bit of repeated data, but as long as there is a common way to retrieve PLIST and PROMOPACK data sources, i'm all for it

**Comment by peardox** [comment]
> imo it's a bit of repeated data, but as long as there is a common way to retrieve PLIST and PROMOPACK data sources, i'm all for it

What data is repeated?

It's not like I'm suggesting sticking a repeated currency field in the middle of a load of numerical information

**Comment by kodawah** [comment]
> What data is repeated?

Each PLIST-ZNR, PLIST-KHM etc will contain the full set of cards with all the data, which will be identical to the ones present in the PLIST set. My suggestion was aiming at reducing the repetition and just add the relevant information (aka in which of the many PLIST sub sets a card is present)

> It's not like I'm suggesting sticking a repeated currency field in the middle of a load of numerical information

this is extremely passive aggressive and really unproductive, if you are grumpy about something it's ok to walk away from the keyboard and come back when you have something meaningful to say

**Comment by axxroytovu** [comment]
We have the "boosters/sheets" for other sets, why couldn't we use the same formatting for the list? That would also allow us to include relative rarities and weights of specific cards, like the universes within or other rare/common distinctions.

Formatting would look like:

PLIST.json
```
{
  "data": {
    "booster": {
      "ZNR": {
        "boosters": [{"contents": {"znr_list_common": 1}, "weight":3}, {"contents": {"znr_list_rare": 1}, "weight":1], "boostersTotalWeight": 4, "sheets":{
          "znr_list_common": {"cards": {"d16b7c1d-e83f-5a84-8554-5742dd26ecc0": 1, ...}},
          "znr_list_rare": {"cards": {"97d16e5c-50ee-5132-a1c1-bc098c954c51": 1, ...}}
        }
      }
    }
  }
}
```
Because each sheet is independent, the data can be scraped from the magic.wizards.com blog posts. We would need to make some assumptions about how the rare/uncommon cards are distributed in other sets, since I can only really find the relative rarity info for ZNR.

The changes could be made in the MagicSealedData repo, I don't know how that would affect other projects. I'll ask over there.

Thoughts?

**Comment by axxroytovu** [comment]
List sheets are available in each set json, starting with ONE. Other sets are incoming.

ONE.JSON:
`data/booster/set/sheets/theList`

**Comment by axxroytovu** [comment]
This should be resolved for all current sets, with each set booster listing its own valid cards. Let me know if there are any other changes needed to the methodology. 

Additionally, you can search for The List cards in mtg.wtf using queries like [sheet:plist/znr](https://mtg.wtf/card?q=sheet%3Aplist%2Fznr).

**Comment by ZeldaZach** [comment]
Since we now have the data logged in SET[data][sheets][theList], I would call this issue resolved :)

**Comment by kodawah** [comment]
shouldn't we use this information to fill in the `subsets` property? or is there a dependency problem

**Comment by axxroytovu** [comment]
I don't think there's a dependency problem. Let me look into it.
